src/aa_device_spectral_whitening_kernel.cu

```diff
@@ -0,0 +1,214 @@
+		B_M = aa_shfl_down(AA_ASSUME_MASK, (*M), q);
+		B_S = aa_shfl_down(AA_ASSUME_MASK, (*S), q);
+		B_j = aa_shfl_down(AA_ASSUME_MASK, (*j), q);
+		
+		if(B_j>0){
+			if( (*j)==0 ) {
+				(*S) = B_S;
+				(*M) = B_M;
+				(*j) = B_j;
+			}
+			else {
+				Merge(M, S, j, B_M, B_S, B_j);
+			}
+		}
+	}
+}
+
+
+__device__ __inline__ void MSD_block(float *d_input, int *x_steps, unsigned long int *nSamples, float *s_par_MSD, int *s_par_nElements, float *mean, float *stdev){	
+	float M, S, ftemp;
+	int j;	
+
+	M=0; S=0; j=0;
+	int spos = threadIdx.x;
+	if( spos < (*nSamples) ){
+		ftemp = (float) d_input[spos];
+		Initiate( &M, &S, &j, ftemp);
+		
+		spos = spos + blockDim.x;
+		for (int xf = 1; xf < (*x_steps); xf++) {
+			if( spos < (*nSamples) ){
+				ftemp = (float) d_input[spos];
+				Add_one( &M, &S, &j, ftemp);
+				spos = spos + blockDim.x;
+			}
+		}
+	}
+	
+	s_par_MSD[threadIdx.x] = M;
+	s_par_MSD[128 + threadIdx.x] = S;
+	s_par_nElements[threadIdx.x] = j;
+	
+	__syncthreads();
+	
+	Reduce_SM( &M, &S, &j, s_par_MSD, s_par_nElements );
+	Reduce_WARP( &M, &S, &j);
+	
+	if(threadIdx.x == 0){
+		(*mean) = M / (double) j;
+		(*stdev) = sqrt(S / (double) j);
+		s_par_MSD[0] = (*mean);
+		s_par_MSD[1] = (*stdev);
+	}
+	
+	__syncthreads();
+	
+	(*mean) = s_par_MSD[0];
+	(*stdev) = s_par_MSD[1];
+	
+	__syncthreads();
+}
+
+
+__device__ __inline__ void MSD_block_outlier_rejection(float *d_input, int *x_steps, unsigned long int *nSamples, float *s_par_MSD, int *s_par_nElements, float *mean, float *stdev){	
+	float M, S, ftemp;
+	int j;
+	
+	float sigma = 3.0;
+	float limit_down = (*mean) - sigma*(*stdev);
+	float limit_up = (*mean) + sigma*(*stdev);
+
+	M=0;	S=0;	j=0;
+	int spos = threadIdx.x;
+	if( spos < (*nSamples) ){
+		for (int xf = 0; xf < (*x_steps); xf++) {
+			if( spos < (*nSamples) ){
+				ftemp = (float) d_input[spos];
+				if( (ftemp>limit_down) && (ftemp < limit_up) ){
+					if(j==0){
+						Initiate( &M, &S, &j, ftemp);
+					}
+					else{
+						Add_one( &M, &S, &j, ftemp);
+					}			
+				}
+				spos = spos + blockDim.x;
+			}
+		}	
+	}
+
+	s_par_MSD[threadIdx.x] = M;
+	s_par_MSD[blockDim.x + threadIdx.x] = S;
+	s_par_nElements[threadIdx.x] = j;
+	
+	__syncthreads();
+	
+	Reduce_SM( &M, &S, &j, s_par_MSD, s_par_nElements );
+	Reduce_WARP( &M, &S, &j);
+	
+	if(threadIdx.x == 0){
+		(*mean) = M / (double) j;
+		(*stdev) = sqrt(S / (double) j);
+		s_par_MSD[0] = (*mean);
+		s_par_MSD[1] = (*stdev);
+	}
+	
+	__syncthreads();
+	
+	(*mean) = s_par_MSD[0];
+	(*stdev) = s_par_MSD[1];
+	
+	__syncthreads();
+}
+
+
+__global__ void GPU_kernel_spectrum_whitening_SGP1(float *d_input, unsigned long int nSamples) {
+	__shared__ float s_par_MSD[256];
+	__shared__ int s_par_nElements[128];
+
+	float mean, stdev;
+	unsigned long int gpos = blockIdx.y*nSamples;
+	unsigned long int spos = blockIdx.x*blockDim.x;
+	int x_steps = 1;
+	
+	MSD_block(&d_input[gpos + spos], &x_steps, &nSamples, s_par_MSD, s_par_nElements, &mean, &stdev);
+	
```

Comment:
Just wondering is it still in the memory size of d_input (mean the gpos + spos)
	dim3 grid_size((nSamples + 128 - 1)/128, nDMs , 1);
	dim3 block_size(128, 1, 1);

Reply:
This should be fine. gpos moves the coordinate to the start of the DM-trial while spos is moving coordinates within the DM-trial so gpos is y-axis and spos is x-axis. Besides there is an if branch within the MSD_block function.

Reply:
I missed the if branch probably.